cpp/src/arrow/io/interfaces.cc

```diff
@@ -70,5 +74,55 @@ Status Writable::Write(const std::string& data) {
 
 Status Writable::Flush() { return Status::OK(); }
 
+class FileSegmentReader : public InputStream {
+ public:
+  FileSegmentReader(std::shared_ptr<RandomAccessFile> file, int64_t file_offset,
+                    int64_t nbytes)
+      : file_(std::move(file)), position_(0), file_offset_(file_offset), nbytes_(nbytes) {
+    FileInterface::set_mode(FileMode::READ);
+  }
+
+  Status Close() override { return Status::OK(); }
```

Comment:
Not required, but for sanity I prefer if `Close()` sets a flag that then prevents from calling other methods.

Reply:
Sure, I'll add that.